src/tables/colr.rs

```diff
@@ -0,0 +1,122 @@
+
+use crate::cpal::{self, Color};
+use crate::parser::{FromData, LazyArray16, Offset, Offset32, Stream};
+use crate::GlyphId;
+
+/// A [base glyph](https://learn.microsoft.com/en-us/typography/opentype/spec/colr#baseglyph-and-layer-records).
+#[derive(Clone, Copy, Debug)]
+struct BaseGlyphRecord {
+    glyph_id: GlyphId,
+    first_layer_index: u16,
+    num_layers: u16,
+}
+
+impl FromData for BaseGlyphRecord {
+    const SIZE: usize = 6;
+
+    fn parse(data: &[u8]) -> Option<Self> {
+        let mut s = Stream::new(data);
+        Some(Self {
+            glyph_id: s.read::<GlyphId>()?,
+            first_layer_index: s.read::<u16>()?,
+            num_layers: s.read::<u16>()?,
+        })
+    }
+}
+
+/// A [layer](https://learn.microsoft.com/en-us/typography/opentype/spec/colr#baseglyph-and-layer-records).
+#[derive(Clone, Copy, Debug)]
+struct LayerRecord {
+    glyph_id: GlyphId,
+    palette_index: u16,
+}
+
+impl FromData for LayerRecord {
+    const SIZE: usize = 4;
+
+    fn parse(data: &[u8]) -> Option<Self> {
+        let mut s = Stream::new(data);
+        Some(Self {
+            glyph_id: s.read::<GlyphId>()?,
+            palette_index: s.read::<u16>()?,
+        })
+    }
+}
+
+/// A trait for color glyph painting.
+pub trait ColorGlyphPainter {
+    /// Paints an outline glyph in the given color.
+    fn glyph(&mut self, id: GlyphId, color: Color);
+}
+
+/// A [Color Table](
+/// https://docs.microsoft.com/en-us/typography/opentype/spec/colr).
+///
+/// Currently, only version 0 is supported.
+#[derive(Clone, Copy, Debug)]
+pub struct Table<'a> {
+    palettes: cpal::Table<'a>,
+    base_glyphs: LazyArray16<'a, BaseGlyphRecord>,
+    layers: LazyArray16<'a, LayerRecord>,
+}
+
+impl<'a> Table<'a> {
+    /// Parses a table from raw data.
+    pub fn parse(palettes: cpal::Table<'a>, data: &'a [u8]) -> Option<Self> {
+        let mut s = Stream::new(data);
+
+        let version = s.read::<u16>()?;
+        if version > 1 {
+            return None;
+        }
+
+        let num_base_glyphs = s.read::<u16>()?;
+        let base_glyphs_offset = s.read::<Offset32>()?;
+        let layers_offset = s.read::<Offset32>()?;
+        let num_layers = s.read::<u16>()?;
+
+        let base_glyphs = Stream::new_at(data, base_glyphs_offset.to_usize())?
+            .read_array16::<BaseGlyphRecord>(num_base_glyphs)?;
+
+        let layers = Stream::new_at(data, layers_offset.to_usize())?
+            .read_array16::<LayerRecord>(num_layers)?;
+
+        Some(Self {
+            palettes,
+            base_glyphs,
+            layers,
+        })
+    }
+
+    /// Whether the table contains a definition for the given glyph.
+    pub fn contains(&self, glyph_id: GlyphId) -> bool {
+        self.base_glyphs
+            .binary_search_by(|base| base.glyph_id.cmp(&glyph_id))
+            .is_some()
+    }
+
+    /// Paints the color glyph.
+    pub fn paint(
+        &self,
+        glyph_id: GlyphId,
+        palette: u16,
```

Comment:
@laurmaedje From where the `palette` values comes from?

Reply:
The CPAL table has one or multiple palettes and this value selects which one. 0 will always work. I should probably have added a function to CPAL that gives you the number of palettes. It's not too useful without the palette names though and this is a CPALv1 feature I intentionally skipped.

This value would only be something other than zero if an application can make sense of multiple color palettes in one font, e.g. by exposing it to the user or picking a darker palette in dark mode.

Reply:
We're talking about `CPAL.numPalettes` right? I think I get it now. I will update the docs and the API.

PS: Currently trying to add COLR support to `font2svg` example. Seems to be working fine. But we do have to provide some API to check if the outline is just "a layer" and not a stand-alone glyph.

Reply:
> We're talking about `CPAL.numPalettes` right? I think I get it now. I will update the docs and the API.

Yes, numPalettes is what I meant.

> But we do have to provide some API to check if the outline is just "a layer" and not a stand-alone glyph.

I think technically a glyph could be both. For the purposes of the example, you could maybe use the cmap.codepoints API in combination with glyph_index to find all glyphs that are referenced by the cmap table.

Reply:
As always, everything is far more complicated than it should be.

1. I still have no idea how to check if a glyph is a base glyph,  a color layer glyph or just a normal outline. Checking all layers in COLR would be very expensive.
2. We have to somehow handle 0xFFFF palette. "A paletteIndex value of 0xFFFF is a special case, indicating that the text foreground color (as determined by the application) is to be used." Meaning that even for v0 we have to have `Painter::foreground`

Reply:
1. Why do you want to check that in the first place? I think it's not designed to be checked for.

2. Oh, that's my bad. I think the painter callback could just give an `Option<Color>` where `None` means foreground?

Reply:
1. Mainly for `font2svg`. But I guess in the normal use case it wouldn't be needed.

Reply:
For font2svg it's maybe not that big of a deal if it's expensive.